File: json-tools/scripts/merge.py
#!/usr/bin/env python3
# Copyright 2026 Yevgeniy Zaremba
# SPDX-License-Identifier: Apache-2.0

"""Merge or concatenate JSON files."""

from __future__ import annotations

import argparse
import copy
from typing import Any

from common import first_value, load_json, write_json


def shallow_merge(objs: list[dict[str, Any]]) -> dict[str, Any]:
    out: dict[str, Any] = {}
    for obj in objs:
        out.update(obj)
    return out


def deep_merge_values(left: Any, right: Any) -> Any:
    if isinstance(left, dict) and isinstance(right, dict):
        merged = dict(left)
        for key, value in right.items():
            if key in merged:
                merged[key] = deep_merge_values(merged[key], value)
            else:
                merged[key] = copy.deepcopy(value)
        return merged
    if isinstance(left, list) and isinstance(right, list):
        return left + right
    return copy.deepcopy(right)


def merge_arrays(arrays: list[list[Any]], unique_by: str | None) -> list[Any]:
    combined: list[Any] = []
    for arr in arrays:
        combined.extend(arr)
    if not unique_by:
        return combined
    seen: set[str] = set()
    deduped: list[Any] = []
    for item in combined:
        key = first_value(item, unique_by)
        token = repr(key)
        if token in seen:
            continue
        seen.add(token)
        deduped.append(item)
    return deduped


def main() -> None:
    parser = argparse.ArgumentParser(description="Merge multiple JSON files.")
    parser.add_argument("inputs", nargs="+", help="Input JSON files in merge order.")
    parser.add_argument(
        "--mode",
        choices=["concat", "shallow", "deep"],
        default="concat",
        help="concat expects arrays; shallow/deep expect objects.",
    )
    parser.add_argument("--unique-by", help="Field path used to deduplicate in concat mode.")
    parser.add_argument("--compact", action="store_true", help="Emit compact JSON output.")
    args = parser.parse_args()

    docs = [load_json(path) for path in args.inputs]
    if args.mode == "concat":
        arrays = [doc for doc in docs if isinstance(doc, list)]
        result = merge_arrays(arrays, args.unique_by)
    elif args.mode == "shallow":
        objects = [doc for doc in docs if isinstance(doc, dict)]
        result = shallow_merge(objects)
    else:
        objects = [doc for doc in docs if isinstance(doc, dict)]
        result: dict[str, Any] = {}
        for obj in objects:
            result = deep_merge_values(result, obj)
    write_json(result, compact=args.compact)


if __name__ == "__main__":
    main()

File: json-tools/scripts-node/merge.mjs
#!/usr/bin/env node
// Copyright 2026 Yevgeniy Zaremba
// SPDX-License-Identifier: Apache-2.0

import { parseArgs } from "node:util";
import { firstValue, loadJson, writeJson } from "./common.mjs";

function shallowMerge(objects) {
  const out = {};
  for (const obj of objects) {
    Object.assign(out, obj);
  }
  return out;
}

function deepClone(value) {
  if (value && typeof value === "object") {
    return JSON.parse(JSON.stringify(value));
  }
  return value;
}

function deepMergeValues(left, right) {
  if (left && typeof left === "object" && !Array.isArray(left)
    && right && typeof right === "object" && !Array.isArray(right)) {
    const merged = { ...left };
    for (const [key, value] of Object.entries(right)) {
      if (Object.hasOwn(merged, key)) {
        merged[key] = deepMergeValues(merged[key], value);
      } else {
        merged[key] = deepClone(value);
      }
    }
    return merged;
  }
  if (Array.isArray(left) && Array.isArray(right)) {
    return [...left, ...right];
  }
  return deepClone(right);
}

function mergeArrays(arrays, uniqueBy) {
  const combined = [];
  for (const arr of arrays) {
    combined.push(...arr);
  }
  if (!uniqueBy) {
    return combined;
  }
  const seen = new Set();
  const deduped = [];
  for (const item of combined) {
    const key = firstValue(item, uniqueBy);
    const token = `${typeof key}:${JSON.stringify(key)}`;
    if (seen.has(token)) {
      continue;
    }
    seen.add(token);
    deduped.push(item);
  }
  return deduped;
}

function main() {
  const { values, positionals } = parseArgs({
    options: {
      mode: { type: "string", default: "concat" },
      "unique-by": { type: "string" },
      compact: { type: "boolean", default: false },
    },
    allowPositionals: true,
  });

  if (positionals.length < 1) {
    throw new Error("At least one input path is required.");
  }
  if (!["concat", "shallow", "deep"].includes(values.mode)) {
    throw new Error("--mode must be one of: concat, shallow, deep");
  }

  const docs = positionals.map((path) => loadJson(path));
  let result;
  if (values.mode === "concat") {
    const arrays = docs.filter((doc) => Array.isArray(doc));
    result = mergeArrays(arrays, values["unique-by"]);
  } else if (values.mode === "shallow") {
    const objects = docs.filter((doc) => doc && typeof doc === "object" && !Array.isArray(doc));
    result = shallowMerge(objects);
  } else {
    const objects = docs.filter((doc) => doc && typeof doc === "object" && !Array.isArray(doc));
    result = {};
    for (const obj of objects) {
      result = deepMergeValues(result, obj);
    }
  }
  writeJson(result, values.compact);
}

main();
